// @jest/expect/build/dist/index.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.jestExpect = void 0;

function _expect() {
  var data = require('expect');

  _expect = function _expect() {
    return data;
  };

  return data;
}

function _jestSnapshot() {
  var data = require('jest-snapshot');

  _jestSnapshot = function _jestSnapshot() {
    return data;
  };

  return data;
}
/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */


function createJestExpect() {
  _expect().expect.extend({
    toMatchInlineSnapshot: _jestSnapshot().toMatchInlineSnapshot,
    toMatchSnapshot: _jestSnapshot().toMatchSnapshot,
    toThrowErrorMatchingInlineSnapshot: _jestSnapshot().toThrowErrorMatchingInlineSnapshot,
    toThrowErrorMatchingSnapshot: _jestSnapshot().toThrowErrorMatchingSnapshot
  });

  _expect().expect.addSnapshotSerializer = _jestSnapshot().addSerializer;
  return _expect().expect;
}

var jestExpect = createJestExpect();
exports.jestExpect = jestExpect;
