// @babel/traverse/lib/path/lib/dist/virtual-types.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Var = exports.User = exports.Statement = exports.SpreadProperty = exports.Scope = exports.RestProperty = exports.ReferencedMemberExpression = exports.ReferencedIdentifier = exports.Referenced = exports.Pure = exports.NumericLiteralTypeAnnotation = exports.Generated = exports.ForAwaitStatement = exports.Flow = exports.Expression = exports.ExistentialTypeParam = exports.BlockScoped = exports.BindingIdentifier = void 0;
var ReferencedIdentifier = ["Identifier", "JSXIdentifier"];
exports.ReferencedIdentifier = ReferencedIdentifier;
var ReferencedMemberExpression = ["MemberExpression"];
exports.ReferencedMemberExpression = ReferencedMemberExpression;
var BindingIdentifier = ["Identifier"];
exports.BindingIdentifier = BindingIdentifier;
var Statement = ["Statement"];
exports.Statement = Statement;
var Expression = ["Expression"];
exports.Expression = Expression;
var Scope = ["Scopable", "Pattern"];
exports.Scope = Scope;
var Referenced = null;
exports.Referenced = Referenced;
var BlockScoped = null;
exports.BlockScoped = BlockScoped;
var Var = ["VariableDeclaration"];
exports.Var = Var;
var User = null;
exports.User = User;
var Generated = null;
exports.Generated = Generated;
var Pure = null;
exports.Pure = Pure;
var Flow = ["Flow", "ImportDeclaration", "ExportDeclaration", "ImportSpecifier"];
exports.Flow = Flow;
var RestProperty = ["RestElement"];
exports.RestProperty = RestProperty;
var SpreadProperty = ["RestElement"];
exports.SpreadProperty = SpreadProperty;
var ExistentialTypeParam = ["ExistsTypeAnnotation"];
exports.ExistentialTypeParam = ExistentialTypeParam;
var NumericLiteralTypeAnnotation = ["NumberLiteralTypeAnnotation"];
exports.NumericLiteralTypeAnnotation = NumericLiteralTypeAnnotation;
var ForAwaitStatement = ["ForOfStatement"];
exports.ForAwaitStatement = ForAwaitStatement;
