// yargs-parser/dist/browser.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _stringUtils = require("./build/lib/string-utils.js");

var _yargsParser = require("./build/lib/yargs-parser.js");

// Main entrypoint for ESM web browser environments. Avoids using Node.js
// specific libraries, such as "path".
//
// TODO: figure out reasonable web equivalents for "resolve", "normalize", etc.
var parser = new _yargsParser.YargsParser({
  cwd: function cwd() {
    return '';
  },
  format: function format(str, arg) {
    return str.replace('%s', arg);
  },
  normalize: function normalize(str) {
    return str;
  },
  resolve: function resolve(str) {
    return str;
  },
  require: function require() {
    throw Error('loading config from files not currently supported in browser');
  },
  env: function env() {}
});

var yargsParser = function Parser(args, opts) {
  var result = parser.parse(args.slice(), opts);
  return result.argv;
};

yargsParser.detailed = function (args, opts) {
  return parser.parse(args.slice(), opts);
};

yargsParser.camelCase = _stringUtils.camelCase;
yargsParser.decamelize = _stringUtils.decamelize;
yargsParser.looksLikeNumber = _stringUtils.looksLikeNumber;
var _default = yargsParser;
exports["default"] = _default;
