// istanbul-lib-instrument/src/dist/read-coverage.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var _require = require('@babel/core'),
    parseSync = _require.parseSync,
    traverse = _require.traverse;

var _require2 = require('@istanbuljs/schema'),
    defaults = _require2.defaults;

var _require3 = require('./constants'),
    MAGIC_KEY = _require3.MAGIC_KEY,
    MAGIC_VALUE = _require3.MAGIC_VALUE;

function getAst(code) {
  if (_typeof(code) === 'object' && typeof code.type === 'string') {
    // Assume code is already a babel ast.
    return code;
  }

  if (typeof code !== 'string') {
    throw new Error('Code must be a string');
  } // Parse as leniently as possible


  return parseSync(code, {
    babelrc: false,
    configFile: false,
    parserOpts: {
      allowAwaitOutsideFunction: true,
      allowImportExportEverywhere: true,
      allowReturnOutsideFunction: true,
      allowSuperOutsideMethod: true,
      sourceType: 'script',
      plugins: defaults.instrumenter.parserPlugins
    }
  });
}

module.exports = function readInitialCoverage(code) {
  var ast = getAst(code);
  var covScope;
  traverse(ast, {
    ObjectProperty: function ObjectProperty(path) {
      var node = path.node;

      if (!node.computed && path.get('key').isIdentifier() && node.key.name === MAGIC_KEY) {
        var magicValue = path.get('value').evaluate();

        if (!magicValue.confident || magicValue.value !== MAGIC_VALUE) {
          return;
        }

        covScope = path.scope.getFunctionParent() || path.scope.getProgramParent();
        path.stop();
      }
    }
  });

  if (!covScope) {
    return null;
  }

  var result = {};

  for (var _i = 0, _arr = ['path', 'hash', 'gcv', 'coverageData']; _i < _arr.length; _i++) {
    var key = _arr[_i];
    var binding = covScope.getOwnBinding(key);

    if (!binding) {
      return null;
    }

    var valuePath = binding.path.get('init');
    var value = valuePath.evaluate();

    if (!value.confident) {
      return null;
    }

    result[key] = value.value;
  }

  delete result.coverageData[MAGIC_KEY];
  delete result.coverageData.hash;
  return result;
};
