// jest-each/build/table/dist/template.dev.js
'use strict';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports["default"] = template;

var _interpolation = require('./interpolation');
/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */


function template(title, headings, row) {
  var table = convertRowToTable(row, headings);
  var templates = convertTableToTemplates(table, headings);
  return templates.map(function (template, index) {
    return {
      arguments: [template],
      title: (0, _interpolation.interpolateVariables)(title, template, index)
    };
  });
}

var convertRowToTable = function convertRowToTable(row, headings) {
  return Array.from({
    length: row.length / headings.length
  }).map(function (_, index) {
    return row.slice(index * headings.length, index * headings.length + headings.length);
  });
};

var convertTableToTemplates = function convertTableToTemplates(table, headings) {
  return table.map(function (row) {
    return row.reduce(function (acc, value, index) {
      return Object.assign(acc, _defineProperty({}, headings[index], value));
    }, {});
  });
};
