// jest-config/build/dist/utils.dev.js
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports.resolve = exports.replaceRootDirInPath = exports.isJSONString = exports.escapeGlobCharacters = exports._replaceRootDirTags = exports.DOCUMENTATION_NOTE = exports.BULLET = void 0;

function path() {
  var data = _interopRequireWildcard(require('path'));

  path = function path() {
    return data;
  };

  return data;
}

function _chalk() {
  var data = _interopRequireDefault(require('chalk'));

  _chalk = function _chalk() {
    return data;
  };

  return data;
}

function _jestResolve() {
  var data = _interopRequireDefault(require('jest-resolve'));

  _jestResolve = function _jestResolve() {
    return data;
  };

  return data;
}

function _jestValidate() {
  var data = require('jest-validate');

  _jestValidate = function _jestValidate() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== 'function') return null;
  var cacheBabelInterop = new WeakMap();
  var cacheNodeInterop = new WeakMap();
  return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}

function _interopRequireWildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }

  if (obj === null || _typeof(obj) !== 'object' && typeof obj !== 'function') {
    return {
      "default": obj
    };
  }

  var cache = _getRequireWildcardCache(nodeInterop);

  if (cache && cache.has(obj)) {
    return cache.get(obj);
  }

  var newObj = {};
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;

  for (var key in obj) {
    if (key !== 'default' && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;

      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }

  newObj["default"] = obj;

  if (cache) {
    cache.set(obj, newObj);
  }

  return newObj;
}
/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */


var BULLET = _chalk()["default"].bold("\u25CF ");

exports.BULLET = BULLET;
var DOCUMENTATION_NOTE = "  ".concat(_chalk()["default"].bold('Configuration Documentation:'), "\n  https://jestjs.io/docs/configuration\n");
exports.DOCUMENTATION_NOTE = DOCUMENTATION_NOTE;

var createValidationError = function createValidationError(message) {
  return new (_jestValidate().ValidationError)("".concat(BULLET, "Validation Error"), message, DOCUMENTATION_NOTE);
};

var resolve = function resolve(resolver, _ref) {
  var key = _ref.key,
      filePath = _ref.filePath,
      rootDir = _ref.rootDir,
      optional = _ref.optional;

  var module = _jestResolve()["default"].findNodeModule(replaceRootDirInPath(rootDir, filePath), {
    basedir: rootDir,
    resolver: resolver || undefined
  });

  if (!module && !optional) {
    throw createValidationError("  Module ".concat(_chalk()["default"].bold(filePath), " in the ").concat(_chalk()["default"].bold(key), " option was not found.\n         ").concat(_chalk()["default"].bold('<rootDir>'), " is: ").concat(rootDir));
  } /// can cast as string since nulls will be thrown


  return module;
};

exports.resolve = resolve;

var escapeGlobCharacters = function escapeGlobCharacters(path) {
  return path.replace(/([()*{}[\]!?\\])/g, '\\$1');
};

exports.escapeGlobCharacters = escapeGlobCharacters;

var replaceRootDirInPath = function replaceRootDirInPath(rootDir, filePath) {
  if (!/^<rootDir>/.test(filePath)) {
    return filePath;
  }

  return path().resolve(rootDir, path().normalize("./".concat(filePath.substring('<rootDir>'.length))));
};

exports.replaceRootDirInPath = replaceRootDirInPath;

var _replaceRootDirInObject = function _replaceRootDirInObject(rootDir, config) {
  var newConfig = {};

  for (var configKey in config) {
    newConfig[configKey] = configKey === 'rootDir' ? config[configKey] : _replaceRootDirTags(rootDir, config[configKey]);
  }

  return newConfig;
};

var _replaceRootDirTags = function _replaceRootDirTags(rootDir, config) {
  if (config == null) {
    return config;
  }

  switch (_typeof(config)) {
    case 'object':
      if (Array.isArray(config)) {
        /// can be string[] or {}[]
        return config.map(function (item) {
          return _replaceRootDirTags(rootDir, item);
        });
      }

      if (config instanceof RegExp) {
        return config;
      }

      return _replaceRootDirInObject(rootDir, config);

    case 'string':
      return replaceRootDirInPath(rootDir, config);
  }

  return config;
};

exports._replaceRootDirTags = _replaceRootDirTags; // newtype

var isJSONString = function isJSONString(text) {
  return text != null && typeof text === 'string' && text.startsWith('{') && text.endsWith('}');
};

exports.isJSONString = isJSONString;
