// @babel/core/lib/config/dist/resolve-targets.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.resolveBrowserslistConfigFile = resolveBrowserslistConfigFile;
exports.resolveTargets = resolveTargets;

function _path() {
  var data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

function _helperCompilationTargets() {
  var data = require("@babel/helper-compilation-targets");

  _helperCompilationTargets = function _helperCompilationTargets() {
    return data;
  };

  return data;
}

({});

function resolveBrowserslistConfigFile(browserslistConfigFile, configFileDir) {
  return _path().resolve(configFileDir, browserslistConfigFile);
}

function resolveTargets(options, root) {
  var optTargets = options.targets;
  var targets;

  if (typeof optTargets === "string" || Array.isArray(optTargets)) {
    targets = {
      browsers: optTargets
    };
  } else if (optTargets) {
    if ("esmodules" in optTargets) {
      targets = Object.assign({}, optTargets, {
        esmodules: "intersect"
      });
    } else {
      targets = optTargets;
    }
  }

  var browserslistConfigFile = options.browserslistConfigFile;
  var configFile;
  var ignoreBrowserslistConfig = false;

  if (typeof browserslistConfigFile === "string") {
    configFile = browserslistConfigFile;
  } else {
    ignoreBrowserslistConfig = browserslistConfigFile === false;
  }

  return (0, _helperCompilationTargets()["default"])(targets, {
    ignoreBrowserslistConfig: ignoreBrowserslistConfig,
    configFile: configFile,
    configPath: root,
    browserslistEnv: options.browserslistEnv
  });
}

0 && 0;
